### ISSUE TITLE
Should GAP code assume memory is always allocated in word-sized blocks?

### ISSUE THREAD

**Issue by ChrisJefferson** [gap-system/gap#2619]
It turns out there are a few places in GAP, either accidentally or on purpose, which assume that the size of an allocated block of memory is always a multiple of the size of void*.

For example: `intfuncs.c:104` reads a block of memory as uint4s.

In GAP at present, this always works, as we always round up memory allocations. I found this because I added support to catch this kind of bug.

I believe this is a bug -- if code calls NewBag(.., X), then it should only assume it can read & write X bytes, not X rounded up to the nearest multiple of 4 or 8. However, I wanted to ask if there was general agreement.

**Comment by stevelinton** [comment]
Agreed. If we wanted the other behaviour, we should specify the size in words. We do, however promise that allocated bags are word-aligned I think.

**Comment by fingolfin** [comment]
I am not sure I agree. Rounding bag sizes to full words allows to simplify a lot of code, so it's useful. And I see no drawback in just making this behavior official. Most memory allocators do that internally anyway.

**Comment by ChrisJefferson** [comment]
I'm fairly neutral on this. It actually seems to be less help -- most other uses (except this use in random) seems to be users writing manual memcpy loops in 4 byte steps, when they could be just using memcpy (where we may as well pass the exact number of bytes).

I can continue and see how big a patch would be (I think using memcpy is better anyway).

**Comment by ChrisJefferson** [comment]
There are quite a few places where we are assuming word aligned. These can be broadly divided (mostly) into two types:

1) The bag copying functions use do a word-based loop. These probably should be replaced by memcpy, at which point we could either explicitly continue copying based on words, or switch to only copying the bag.

2) Various functions (in particular the vec8 ones) work on word-based chunks. I haven't worked through how hard they would be to fix.

If we keep the extra bytes readable/writable, we could enforce these extra bytes are 0, although we could only reasonably check that at GC time, or allow them to take any value (in which case we should be more careful about ensuring they are always copied correctly).
